sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -836,17 +836,29 @@ class Analyzer(
           Generate(newG.asInstanceOf[Generator], join, outer, qualifier, output, child)
         }
 
+      // If grouping keys have unresolved expressions, we need to replace them with resolved one
+      // in SELECT clauses.
+      case agg @ Aggregate(groups, aggs, child)
+          if child.resolved && aggs.forall(_.resolved) && groups.exists(!_.resolved) =>
+        agg.copy(groupingExpressions = groups.map {
+            case u: UnresolvedAttribute =>
+              aggs.find(ne => resolver(ne.name, u.name)).map {
+                case alias @ Alias(e, _) => e
+                case e => e
+              }.getOrElse(u)
+            case e => e
+          })
+
```

Comment:
@maropu I did not suggest to merge the two code blocks. I suggested to move them next to each other so readers will easily see the code handling Aggregate with two different patterns.